Slow-start batch pod creation of rs, rc, ds, jobs

joelsmith opened the pull request:
Prevent too-large replicas from generating enormous numbers
of events by creating only a few pods at a time, then increasing
the batch size when pod creations succeed. Stop creating batches
of pods when any pod creation errors are encountered.

Todo:

- [x] Add automated tests
- [x] Test ds

Fixes https://github.com/kubernetes/kubernetes/issues/49145

**Release note**:
```release-note
controllers backoff better in face of quota denial
```

joelsmith committed:
Slow-start batch pod creation of rs, rc, ds, jobs

Prevent too-large replicas from generating enormous numbers
of events by creating only a few pods at a time, then increasing
the batch size when pod creations succeed. Stop creating batches
of pods when any pod creation errors are encountered.

joelsmith commented:
cc @gmarek @derekwaynecarr

derekwaynecarr commented:
/ok-to-test

@gmarek -- this is one of the scale problems i discussed with you via slack that we have hit in our clusters.  in general, i think we should call this out as a pattern we should follow for any controller that spawns resources that could be controlled via quota.

derekwaynecarr commented:
@joelsmith -- please create an issue that describes the problem (controllers when denied by quota make too many requests).  link to that issue in your PR description with a Fixes (insert link to issue).  Thanks!

krzyzacy commented:
/test pull-kubernetes-e2e-gce-etcd3

krzyzacy commented:
@kubernetes/test-infra-maintainers seems it happens more often these days?
`
I0718 22:18:56.822] Your "OAuth 2.0 Service Account" credentials are invalid. Please run
I0718 22:18:56.822]   $ gcloud auth login
I0718 22:18:56.822] Type position out of range
`
/test pull-kubernetes-e2e-gce-etcd3

gmarek commented:
I hope that you plan to add tests in this PR:)

joelsmith commented:
It looks like we have some flaky tests. Still getting `Your "OAuth 2.0 Service Account" credentials are invalid.`

/test pull-kubernetes-kubemark-e2e-gce
/test pull-kubernetes-federation-e2e-gce

kargakis commented:
looks good overall

@kubernetes/sig-apps-pr-reviews

aveshagarwal commented:
/test pull-kubernetes-federation-e2e-gce

joelsmith commented:
@gmarek I've added unit tests for each of the 4 controllers.

@derekwaynecarr I've also removed the WIP label now that I've finished verifying the ds behavior on an actual cluster.

I think this PR is ready for review.

kargakis commented:
/approve #49145

This LGTM but I will let @gmarek have a final pass

gmarek commented:
Sorry, I was OOO.

gmarek commented on pkg/controller/daemon/daemoncontroller.go:
```diff
@@ -812,20 +813,42 @@ func (dsc *DaemonSetsController) syncNodes(ds *extensions.DaemonSet, podsToDelet
+	// API service with the pod create requests after one of its pods fails.
+	// Conveniently, this also prevents the event spam that those failures
+	// would generate.
+	for batchSize, pos := 1, 0; createDiff > pos; batchSize, pos = integer.IntMin(2*batchSize, createDiff-(pos+batchSize)), pos+batchSize {
```
This is pretty bad case of code duplication. Please factor out common patterns and refactor controller code if necessary.

gmarek commented:
Idea generally is sound, but I have two comments:
- there's a lot of code duplication which certainly needs to be removed
- we start with batch of size '1', which may be overly cautious. I'd probably start with 10, or some other number, as we generally expect creations to work and don't want to slow them down unnecessarily.

soltysh commented:
I'm going through my backlog of emails and found this. I totally agree with @gmarek here. I'd rather not to force every controller have this in place (as in duplicated) but rather a controller's util function to do so. +1 for the initial batch size, 1 seems to small and with bigger resources this will significantly slow down the setup phase, which in turn might be seen as a regression over previous versions of k8s.

joelsmith commented:
@gmarek , @soltysh I'm not sure that it is a lot of code duplication unless we count the long comments (the long comments that explain the presence of the extra for loop). I'm happy to change it if there really is consensus that it's worth it. Could you take a look at a re-factored alternative that I put on a PR against my own fork? If you like the look of that code, I'll push it into this PR too. I just thought you might be disappointed when you saw how little common slow start code I had actually added here.

The diffs here of the controllers are much more readable when white space is ignored, so if you're not already doing it, I suggest adding `?w=1` to the GH files URL of the PRs. I'll try to add links that contain that parameter here.

[PR with common code factored into a `SlowStart()` function](https://github.com/joelsmith/kubernetes/pull/2/files?w=1)

[The `SlowStart()` function](https://github.com/joelsmith/kubernetes/pull/2/files?w=1#diff-f22f69b57bb9b342baee9943e6143ba0R43)

[This PR's files with `?w=1` added (for comparison)](https://github.com/kubernetes/kubernetes/pull/49142/files?w=1)

If you think I should also try to consolidate the error handling, I can probably pass a few more items to `SlowStart()` (at the expense of making it less generic) to have it do most of the error handling.

I haven't changed the initial starting point yet. @derekwaynecarr asked me to do this work and he's out for a few days, so I think it's probably a good idea to let him chime in too on the specifics.  Just so we're on the same page about what the trade-offs are, starting at 10 instead of 1 provides a speed-up time of (up to) 3 API call round trip times, but causes 10x the API spam and event noise in the error case (since each controller sync iteration will try and fail to start 10 pods instead of 1).

|A  |B |C |
|---|--|--|
|1  |1 |1 |
|2  |2 |1 |
|4  |3 |1 |
|8  |4 |1 |
|15 |4 |2 |
|16 |5 |2 |
|30 |5 |2 |
|31 |5 |3 |
|32 |6 |3 |
|64 |7 |4 |

A: number of pods to start
B: number of API call batches needed for a start value of 1
C: number of API call batches needed for a start value of 10

joelsmith commented:
/retest

joelsmith commented:
@derekwaynecarr Could you take a look and give your opinions? There are 2 open questions here:

1. What is a good starting point for the number of pods to start per iteration? Current code has a start value of 1, and @gmarek proposes we change that to something larger, like 10. My [earlier comment](https://github.com/kubernetes/kubernetes/pull/49142#issuecomment-319580912) shows the tradeoff via a table for those two values.

2. Is it worth refactoring to get rid of duplicate code (much of which is a large comment). The [current version](https://github.com/kubernetes/kubernetes/pull/49142/files?w=1) has the same for loop across all 4 controllers with an enormous comment explaining why it's there. The [alternate version](https://github.com/joelsmith/kubernetes/pull/2/files?w=1) has the for loop and comment moved into a shared utility function. A hypothetical third version could go further by making a `PodSlowStart()` function instead of `SlowStart()` that does more of the common work, but is usable only for batching pod start-ups that fit the pattern.

I'm fine with any approach that we decide is best, so I'm hoping we'll converge quickly on what we want so I can finish up this PR.

Thanks!

kargakis commented:
Drying out the code to avoid duplication can happen in a follow-up PR.

Regarding the pods to start per iteration, does it make sense to expose it
via a flag in the controller manager?

On Fri, Aug 25, 2017 at 9:24 PM, Joel Smith <notifications@github.com>
wrote:

> @derekwaynecarr <https://github.com/derekwaynecarr> Could you take a look
> and give your opinions? There are 2 open questions here:
>
>    1.
>
>    What is a good starting point for the number of pods to start per
>    iteration? Current code has a start value of 1, and @gmarek
>    <https://github.com/gmarek> proposes we change that to something
>    larger, like 10. My earlier comment
>    <https://github.com/kubernetes/kubernetes/pull/49142#issuecomment-319580912>
>    shows the tradeoff via a table for those two values.
>    2.
>
>    Is it worth refactoring to get rid of duplicate code (much of which is
>    a large comment). The current version
>    <https://github.com/kubernetes/kubernetes/pull/49142/files?w=1> has
>    the same for loop across all 4 controllers with an enormous comment
>    explaining why it's there. The alternate version
>    <https://github.com/joelsmith/kubernetes/pull/2/files?w=1> has the for
>    loop and comment moved into a shared utility function. A hypothetical third
>    version could go further by making a PodSlowStart() function instead
>    of SlowStart() that does more of the common work, but is usable only
>    for batching pod start-ups that fit the pattern.
>
> I'm fine with any approach that we decide is best, so I'm hoping we'll
> converge quickly on what we want so I can finish up this PR.
>
> Thanks!
>
> —
> You are receiving this because you are on a team that was mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/49142#issuecomment-325000926>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADuFf9KXp5EiE_fKZyK4-0ziYsZe3NV7ks5sbxFZgaJpZM4Obzmm>
> .
>

derekwaynecarr commented on pkg/controller/daemon/daemon_controller.go:
```diff
@@ -904,31 +905,53 @@ func (dsc *DaemonSetsController) syncNodes(ds *extensions.DaemonSet, podsToDelet
+	// API service with the pod create requests after one of its pods fails.
+	// Conveniently, this also prevents the event spam that those failures
+	// would generate.
+	for batchSize, pos := 1, 0; createDiff > pos; batchSize, pos = integer.IntMin(2*batchSize, createDiff-(pos+batchSize)), pos+batchSize {
```
can you move the initial batch size to a constant that all controllers reference?

that way we can tweak it globally from 1, 2, 4, 8, ... in one place.

derekwaynecarr commented on pkg/controller/job/job_controller.go:
```diff
@@ -620,34 +621,60 @@ func (jm *JobController) manageJob(activePods []*v1.Pod, succeeded int32, job *b
+		// API service with the pod create requests after one of its pods fails.
+		// Conveniently, this also prevents the event spam that those failures
+		// would generate.
+		for batchSize := int32(1); diff > 0; batchSize = integer.Int32Min(2*batchSize, diff) {
```
same here, move initial batch size to a shared constant.

derekwaynecarr commented:
@joelsmith -- can you move the batch size to a global constant so we can then tweak in follow-on to an appropriate value that balances @gmarek concerns?

my opinion is that the potential abuse of a higher initial batch is more concerning than the speed at which we burst from 0-N pods.  i am fine to get a tunable option in controller in future for this exact scenario which would let operators tweak.  i am not going to block this pr on that though as this has been a production issue for us.

we can address the code duplication in a follow-on.

poke when you make that one update and then i will tag.

derekwaynecarr commented:
this is a bug fix i want to ensure makes 1.8 so marking w/ proper milestone.

joelsmith commented:
/retest

joelsmith commented:
@derekwaynecarr  I have made a single constant for the initial batch start and fixed the unit tests accordingly. Rebased, squashed and passing all the tests.

derekwaynecarr commented on pkg/controller/controller_utils.go:
```diff
@@ -65,6 +65,8 @@ const (
 	// 500 pods. Just creation is limited to 20qps, and watching happens with ~10-30s
 	// latency/pod at the scale of 3000 pods over 100 nodes.
 	ExpectationsTimeout = 5 * time.Minute
+	//
```
missing comment text, please describe the field and the impact of increasing the value.

derekwaynecarr commented:
Thanks for the update.

/approve
/lgtm

derekwaynecarr commented:
/test pull-kubernetes-federation-e2e-gce
/test pull-kubernetes-e2e-gce-etcd3

kubemark is unrelated.

joelsmith commented:
Now that the broken test is fixed:
/test pull-kubernetes-kubemark-e2e-gce

joelsmith commented:
looks like a flake
/test pull-kubernetes-e2e-gce-etcd3
